semialign/src/Data/Crosswalk.hs

```diff
@@ -72,18 +79,55 @@ crosswalkVector f = fmap VG.fromList . VG.foldr (alignWith cons . f) nil where
 instance Crosswalk V.Vector where
     crosswalk = crosswalkVector
 
+instance Crosswalk (Either e) where
+    crosswalk _ (Left _) = nil
+    crosswalk f (Right x) = Right <$> f x
+
+instance Crosswalk (These a) where
+    crosswalk _ (This _) = nil
+    crosswalk f (That x) = That <$> f x
+    crosswalk f (These a x) = These a <$> f x
+
 instance Crosswalk ((,) a) where
     crosswalk fun (a, x) = fmap ((,) a) (fun x)
 
 -- can't (shouldn't) do longer tuples until there are Functor and Foldable
 -- instances for them
 
+instance Crosswalk Proxy where
+    crosswalk _ _ = nil
+
+instance Crosswalk (Const r) where
+    crosswalk _ _ = nil
+
+instance (Crosswalk f, Crosswalk g) => Crosswalk (Sum f g) where
+    crosswalk f (InL xs) = InL <$> crosswalk f xs
+    crosswalk f (InR xs) = InR <$> crosswalk f xs
+
+instance (Crosswalk f, Crosswalk g) => Crosswalk (These1 f g) where
+    crosswalk f (This1 xs) = This1 <$> crosswalk f xs
+    crosswalk f (That1 ys) = That1 <$> crosswalk f ys
+    crosswalk f (These1 xs ys) = alignWith go (crosswalk f xs) (crosswalk f ys)
+      where go = these This1 That1 These1
+
 instance (Crosswalk f, Crosswalk g) => Crosswalk (Compose f g) where
     crosswalk f
         = fmap Compose -- can't coerce: maybe the Align-able thing has role nominal
         . crosswalk (crosswalk f)
         . getCompose
 
+data Fill f a = Fill a (f a)
+  deriving (Functor)
+
+instance Align f => Applicative (Fill f) where
+    pure x = Fill x nil
+    Fill deff fs <*> Fill defx xs
+      = Fill (deff defx) (alignWith (uncurry id . fromThese deff defx) fs xs)
+
+instance Traversable t => Crosswalk (MaybeT t) where
```

Comment:
This doesn't look right (even the tests pass). `MaybeT t` is essentially `Compose t Maybe`, and I't expect the instance to be exactly that.

Reply:
If we look at `sequenceL @Maybe` as a sort of generalization of `catMaybes` (that returns Nothing if input was all Nothings), then `Crosswalk (Compose f g)` first removes Nothings from `g`'s, and if that removed all Nothings, it then removes the `g` altogether from `f`, using `Crosswalk f`.

`Crosswalk (MaybeT f)` doesnt do thus, hence being traversable is sufficient.

Reply:
So `Crosswalk (MaybeT t)` behaves differently than `Compose t Maybe`? If it's so, I'd rather not add this instance at all.

Reply:
Yes. It doesn't have to be on the `MaybeT` type constructor specifically (could be a newtype wrapper), but an instance that behaves like this is immensely useful, because it allows implementing `sequenceL` for any type that can implement `catMaybes` (witherable):
```hs
sequenceL @t xs = catMaybes @t . runMaybeT <$> sequenceL @(MaybeT t) (MaybeT (Just <$> xs))
```

Reply:
Are you claiming that every [`Witherable`](https://hackage.haskell.org/package/witherable-0.5/docs/Witherable.html#t:Witherable) is also `Crosswalk`?

Reply:
[Yes](https://github.com/haskellari/these/issues/188)

Reply:
Thoughts?

Reply:
I've removed it from this PR, maybe let's have a separate discussion about it.